tiledb/sm/query/ast/query_ast.h

```diff
@@ -0,0 +1,275 @@
+                 (void*)"" :
+                 condition_value_data_.data()),
+            condition_value_data_.size())
+      , op_(op) {
+    if (condition_value_view_.size() != 0) {
+      memcpy(
+          condition_value_data_.data(), condition_value, condition_value_size);
+    }
+  };
+
+  /** Default destructor. */
+  ~ASTNodeVal() {
+  }
+
+  /** Copy constructor. */
+  ASTNodeVal(const ASTNodeVal& rhs)
+      : field_name_(rhs.field_name_)
+      , condition_value_data_(rhs.condition_value_data_)
+      , condition_value_view_(
+            (rhs.condition_value_view_.content() == nullptr ?
+                 nullptr :
+                 condition_value_data_.data()),
+            condition_value_data_.size())
+      , op_(rhs.op_){};
+
+  /** Move constructor. */
+  ASTNodeVal(ASTNodeVal&& rhs)
+      : field_name_(std::move(rhs.field_name_))
+      , condition_value_data_(std::move(rhs.condition_value_data_))
+      , condition_value_view_(
+            (rhs.condition_value_view_.content() == nullptr ?
+                 nullptr :
+                 condition_value_data_.data()),
+            condition_value_data_.size())
+      , op_(rhs.op_){};
+
+  /** Assignment operator. */
+  ASTNodeVal& operator=(const ASTNodeVal& rhs) {
+    if (this != &rhs) {
+      field_name_ = rhs.field_name_;
+      condition_value_data_ = rhs.condition_value_data_;
+      condition_value_view_ = UntypedDatumView(
+          (rhs.condition_value_view_.content() == nullptr ?
+               nullptr :
+               condition_value_data_.data()),
+          condition_value_data_.size());
+      op_ = rhs.op_;
+    }
+    return *this;
+  }
+
+  /** Move-assignment operator. */
+  ASTNodeVal& operator=(ASTNodeVal&& rhs) {
+    field_name_ = std::move(rhs.field_name_);
+    condition_value_data_ = std::move(rhs.condition_value_data_);
+    condition_value_view_ = std::move(rhs.condition_value_view_);
+    op_ = rhs.op_;
+
+    return *this;
+  }
+
+  /** Returns the tag attached to the node. */
+  ASTNodeTag get_tag() {
+    return tag_;
+  }
+
+  /** Returns a deep copy of the node. */
+  tdb_unique_ptr<ASTNode> clone() {
+    return tdb_unique_ptr<ASTNode>(tdb_new(
+        ASTNodeVal,
+        field_name_,
+        condition_value_data_.data(),
+        condition_value_data_.size(),
+        op_));
+  }
+
+  /** The attribute name. */
+  std::string field_name_;
+
+  /** The value data. */
+  ByteVecValue condition_value_data_;
+
+  /** A view of the value data. */
+  UntypedDatumView condition_value_view_;
+
+  /** The comparison operator. */
+  QueryConditionOp op_;
+
+ private:
+  static const ASTNodeTag tag_{ASTNodeTag::VAL};
+};
+
+class ASTNodeExpr : public ASTNode {
+ public:
+  ASTNodeExpr(
+      std::vector<tdb_unique_ptr<ASTNode>>&& nodes,
+      QueryConditionCombinationOp c_op)
+      : combination_op_(c_op) {
+    for (const auto& elem : nodes) {
+      nodes_.push_back(elem->clone());
```

Comment:
No, we can't because we want a deep copy--we are aiming to have new objects being created when we clone.

Reply:
I am afraid I disagree. Let's take a step back here:
- This is the **constructor** of `ASTNodeExpr`. We must ask ourselves: what do we want to happen to all those `unique_ptr` within `std::vector` that are passed to this constructor ? Since they are unique pointers, they cannot be copied but either _moved_ or just used as read-only accessors to the data they point to. The same holds for `nodes std::vector` that stores those `unique_ptrs`, it can either be moved or passed as `const reference`.
- Here  I believe we want to store nodes as a private variable, so the only way to do this is to actually move the incoming `std::vector` to `nodes_` , transferring in that way the ownership of the incoming `unique_ptrs` to this class.
- The best way to do that is to actually pass `nodes` by value, use `std::move` when passing the vector to the constructor so that an rvalue reference is actually passed (as you already do in this PR). The constructor will then become something like this:
```
  ASTNodeExpr(
      std::vector<tdb_unique_ptr<ASTNode>> nodes,
      QueryConditionCombinationOp c_op)
      : nodes_(std::move(nodes),
        combination_op_(c_op) {};
```
- Note that after calling the constructor we can no longer use the nodes vector on the caller as it will lead to undefined behavior.

Let me know if that's not clear, or let's sync offline.

Reply:
I don't think you can pass a `std::vector<std::unique_ptr<ASTNode>>` by value.  The copy constructor for `std::vector` makes a copy of each of its elements (it is a deep copy).  Since `unique_ptr` can't be copied, a `std::vector<unique_ptr>` can't be copied.

However, there is a question as to why we don't just invoke the move constructor of the vector and do
```c++
ASTNodeExpr( 
  std::vector<tdb_unique_ptr<ASTNode>> nodes&&, 
  QueryConditionCombinationOp c_op) 
  : nodes_(nodes), 
    combination_op_(c_op) {};
```
That would seem to do the Right Thing.

Reply:
This also works and seems to address this concern:

`  ASTNodeExpr(
      std::vector<tdb_unique_ptr<ASTNode>> &&nodes,
      const QueryConditionCombinationOp& c_op)
      : nodes_(std::move(nodes)), combination_op_(c_op) {}`

Reply:
Since `nodes` is already an rvalue reference, you don't need to `std::move` it (`std::move` just turns it's argument into an rvalue reference.

Reply:
Oh, btw:

```
  ASTNodeExpr(
      std::vector<tdb_unique_ptr<ASTNode>> nodes,
      const QueryConditionCombinationOp& c_op)
      : nodes_(std::move(nodes)), combination_op_(c_op) {}
```
This works, but this doesn't work:
```
ASTNodeExpr( 
  std::vector<tdb_unique_ptr<ASTNode>> nodes&&, 
  QueryConditionCombinationOp c_op) 
  : nodes_(nodes), 
    combination_op_(c_op) {};
```